### ISSUE TITLE
`jax.api.mask` Documentation and #292

### ISSUE THREAD

**Issue by tbsexton** [google/jax#2643]
Hi all, I know it's been a while, but I wanted to check on the status of planned "automasking" functionality. This is based on a great conversation with @mattjj in #292 about vmapping over dynamic-length slices of time series observations. 

I see that in the meantime, #1353 has been merged providing an (undocumented) `mask` function. Poking around a bit, the [RNN masking test](https://github.com/google/jax/blob/e9910e0cf11ef6e63c2da2f2f455a3fe030d686d/tests/masking_test.py#L423) looks pretty snazzy! Rather than try to reverse-engineer the behavior, I thought I'd ask here. 

For reference, I did end up re-implementing the example in #292 [using hand-made loops](https://github.com/tbsexton/organizing-tags/blob/master/DEPRECATED/loss.py) in `autograd` and then a (somewhat horrifying) [masking construction](https://github.com/tbsexton/organizing-tags/blob/master/src/models.py) in `pytorch`. 

Needless to say, some nicely behaved masking in the vein of `vmap` would be most welcome!

**Comment by gnecula** [comment]
jax.mask was never finished, we have removed it now, to make room for a more general approach using dynamic shapes.

See #11557
